File: computational-condensed-matter/dwaveDOS_quad.py
import math
import scipy.integrate as inte
import scipy.special as sci
from matplotlib import pyplot as plt

def DOS(stepsize, start, finish, epsilon):
	n = int((finish-start)/stepsize)
	X = []
	N = []
	x = start
	for i in xrange(0,n):
		if abs(x) < 1. and x != 0.:
				X.append(x)
				func = lambda phi: x/math.sqrt(x**2. - math.cos(2.*phi)**2.)
				N.append(inte.quad(func,0.5*math.acos(x)+epsilon,0.5*(math.pi-math.acos(x))-epsilon)[0])
		elif abs(x) > 1:
			X.append(x)
			#N.append(inte.quad(func,0,2.*math.pi)[0])
			N.append((2./math.pi)*sci.ellipk(1./abs(x)))
		x = x + stepsize
	plt.plot(X,N)
	plt.show()
	
DOS(.01,0.,2.,0.001)

File: computational-condensed-matter/dwaveDOS_quad.c
#include <stdio.h>
#include <math.h>
#include <gsl/gsl_integration.h>
#include <gsl/gsl_sf_ellint.h>
#include <mgl2/mgl_cf.h>

#define PI  3.1415926536

double lowlimit(double x, double epsilon)
{
    return (0.5 * acos(x)) + epsilon;
}

double highlimit(double x, double epsilon)
{
    return (0.5 * (PI - acos(x))) - epsilon;
}

double function_to_use(double phi, void *params)
{
    double x = *params;
    double function_to_use = x / sqrt(x**2 - cos(2 * phi)**2
    return function_to_use;
}


double dos(double stepsize, double start, double finish, double epsilon, double eps)
{
    long numsteps = (long) ((finish - start) / stepsize);
    double x[numsteps], f[numsteps];

    gsl_integration_cquad_workspace *w
        = gsl_integration_cquad_workspace_alloc(numsteps);

    double result = 0.0;
    double error = 0.0;
    double alpha = 1.0;
    gsl_function F; /* ADD HERE */
    F.function = &function_to_use;

    long current = start;
    for (long i = 0; i < numsteps; i++) {
        if (abs(current) < 1.0 && current != 0.0) {
            x[i] = current;
            F.params = &current;
            gsl_integration_cquad(&F, lowlimit(current, epsilon), highlimit(current, epsilon)
                0, eps, 1000, w, &result, &error);
            f[i] = result;
        } else if (abs(current) > 1) {
            x[i] = current;
            result = (2.0 / PI) * gsl_sf_ellint_Kcomp(1.0 / abs(current));
            f[i] = result;
        }
        current = current + stepsize;
    }

    mglData xs, fs;
    mglGraph graph;
    xs.Link(x, numsteps);
    fs.Link(f, numsteps);
    graph.Plot(xs, fs);
    graph.WriteFrame("sample.png");
}


int main()
{

    return 0;
}